plugins/docsettingtweak.koplugin/main.lua

```diff
@@ -0,0 +1,106 @@
+local directory_defaults = nil
+local initialized = false
+
+function DocSettingTweak:init()
+    DocSettingTweak:lazyInitialization()
+    self.ui.menu:registerToMainMenu(self)
+end
+
+function DocSettingTweak:loadDefaults()
+    directory_defaults = LuaSettings:open(directory_defaults_path)
+end
+
+function DocSettingTweak:addToMainMenu(menu_items)
+    menu_items.doc_setting_tweak = {
+        text = _("Tweak document settings"),
+        callback = function() DocSettingTweak:editDirectoryDefaults() end,
+    }
+end
+
+function DocSettingTweak:lazyInitialization()
+    if not initialized then
+        if not lfs.attributes(directory_defaults_path, "mode") then
+            FFIUtil.copyFile(FFIUtil.joinPath(self.path, directory_defaults_name),
+                         directory_defaults_path)
+        end
+        initialized = true
+    end
+    DocSettingTweak:loadDefaults()
+end
+
+function DocSettingTweak:editDirectoryDefaults()
+    local directory_defaults_file = io.open(directory_defaults_path, "rb")
+    local defaults = directory_defaults_file:read("*all")
+    directory_defaults_file:close()
+    local config_editor
+    config_editor = InputDialog:new{
+        title = T(_("Directory Defaults: %1"),directory_defaults_path),
+        input = defaults,
+        input_type = "string",
+        fullscreen = true,
+        condensed = true,
+        allow_newline = true,
+        cursor_at_end = false,
+        add_nav_bar = true,
+        reset_callback = function()
+            return defaults
+        end,
+        save_callback = function(content)
+            if content and #content > 0 then
+                local parse_error = util.checkLuaSyntax(content)
+                if not parse_error then
+                    local syntax_okay, syntax_error = pcall(loadstring(content))
+                    if syntax_okay then
+                        directory_defaults_file = io.open(directory_defaults_path, "w")
+                        directory_defaults_file:write(content)
+                        directory_defaults_file:close()
+                        DocSettingTweak:loadDefaults()
+                        return true, _("Defaults saved")
+                    else
+                        return false, T(_("Defaults invalid: %1"), syntax_error)
+                    end
+                else
+                        return false, T(_("Defaults invalid: %1"), parse_error)
+                    end
+            end
+            return false, _("Defaults empty")
+        end,
+    }
+    UIManager:show(config_editor)
+    config_editor:onShowKeyboard()
+end
+
+function DocSettingTweak:onDocSettingsLoad(doc_settings, docfile)
+    if next(doc_settings.data) == nil and directory_defaults.data ~= nil then
+        local base = G_reader_settings:readSetting("home_dir") or filemanagerutil.getDefaultDir()
+        local directory = docfile and docfile ~= "" and FFIUtil.dirname(docfile)
+        while directory:sub(1, #base) == base do
```

Comment:
Can we be called without a `docfile` ?
You take precautions checking if `docfile` is not `nil `and not an empty string to only then get its directory. And if not, directory is `nil`. But then, the `directory:sub()` might crash on a directory that is `nil`.
So, may be just return early if there is no `docfile` provided to make the whole thing simpler clearer.